Add `crate_in_macro_def` lint

smoelius opened the pull request:
This PR adds a lint to check for `crate` as opposed to `$crate` used in a macro definition.

I think this can close #4798. That issue focused on the case where the macro author "imports something into said macro."

But I think use of `crate` is likely to be a bug whether it appears in a `use` statement or not. There could be some use case I am failing to see, though. (cc: @nilscript @flip1995)

changelog: `crate_in_macro_def`

rust-highfive commented:
r? @llogiq

(rust-highfive has picked a reviewer for you, use r? to override)

Alexendoo commented on tests/ui/crate_in_macro_def_allow.rs:
```diff
@@ -0,0 +1,19 @@
+
+#[macro_use]
+mod intentional {
+    // For cases where use of `crate` is intentional, applying `allow` to the macro definition
+    // should suppress the lint.
+    #[allow(clippy::crate_in_macro_def)]
```
This can go in the same test file as above or be removed, as you aren't having to do something specific to get it working

Could be worth noting in the description that the allow attribute has to go on the `macro_rules` itself, ie

```rust
#[allow(clippy::crate_in_macro_def)]
macro_rules! ok { ... crate::foo ... }

macro_rules! lints {
...
    // Still errors since the allow attribute is part of the macro token stream
     #[allow(clippy::crate_in_macro_def)]
    crate::foo
...
}
```

smoelius replied:
> This can go in the same test file as above or be removed

My concern is that someone uses `crate` intentionally, they apply `allow` to the macro definition, and then a future change to Clippy breaks their uses of `allow`. E.g., it would be annoying to have to use `allow` at every call site. The test is meant to help prevent this possibility.

Alexendoo commented on clippy_lints/src/crate_in_macro_def.rs:
````diff
@@ -0,0 +1,100 @@
+    /// definition's crate. Rarely is the former intended. See:
+    /// https://doc.rust-lang.org/reference/macros-by-example.html#hygiene
+    ///
+    /// ### Example
+    /// ```rust
+    /// macro_rules! print_message {
+    ///     () => {
+    ///         println!("{}", crate::MESSAGE);
+    ///     };
+    /// }
+    /// pub const MESSAGE: &str = "Hello!";
+    /// ```
````
For internal macros `crate` is fine, you could check for a `macro_export` attribute using `check_item` to avoid some false positives

smoelius replied:
I am embarrassed that I didn't think of this. :grimacing:

smoelius commented:
Thank you for the review @Alexendoo. I think [cb307bb](https://github.com/rust-lang/rust-clippy/commit/cb307bbfcd123d3bae8f210f2b51eebe68525c27) addresses all of your comments.

llogiq commented on clippy_lints/src/crate_in_macro_def.rs:
````diff
@@ -0,0 +1,126 @@
+    /// ```
+    #[clippy::version = "1.61.0"]
+    pub CRATE_IN_MACRO_DEF,
+    correctness,
````
I don't think this should be a correctness lint. Correctness implies the lint level is `deny`. While the lint will be right in the general case, there may be exceptions and I would not want to stop the build for that. My suggestion here would be to make it a `suspicious` lint (which is `warn` by default).

llogiq commented on clippy_lints/src/crate_in_macro_def.rs:
```diff
@@ -0,0 +1,126 @@
+                    cx,
+                    CRATE_IN_MACRO_DEF,
+                    span,
+                    "reference to the macro call's crate, which is rarely intended",
```
```suggestion
                    "`crate` references the macro call's crate",
```

llogiq commented on clippy_lints/src/crate_in_macro_def.rs:
```diff
@@ -0,0 +1,126 @@
+                    CRATE_IN_MACRO_DEF,
+                    span,
+                    "reference to the macro call's crate, which is rarely intended",
+                    "if reference to the macro definition's crate is intended, use",
```
```suggestion
                    "to reference the macro definition's crate, use",
```

llogiq commented on clippy_lints/src/crate_in_macro_def.rs:
```diff
@@ -0,0 +1,126 @@
+    if_chain! {
+        if let AttrKind::Normal(attr_item, _) = &attr.kind;
+        if let [segment] = attr_item.path.segments.as_slice();
+        if segment.ident.name == sym::macro_export;
+        then {
+            true
+        } else {
+            false
+        }
```
```suggestion
        then {
            segment.ident.name == sym::macro_export
        } else {
            false
        }
```

llogiq reviewed (commented):
This looks good. I have but a small number of suggestions. The only thing that is really worrying me is the lint level.

llogiq commented:
Thank you! @bors r+

bors commented:
:pushpin: Commit d6eb82c0a8d94d21fee110620ee4d0515ac198d3 has been approved by `llogiq`

<!-- @bors r=llogiq d6eb82c0a8d94d21fee110620ee4d0515ac198d3 -->
<!-- homu: {"type":"Approved","sha":"d6eb82c0a8d94d21fee110620ee4d0515ac198d3","approver":"llogiq"} -->

smoelius commented:
@llogiq Sorry, sorry. One more was needed to fix the error message.

llogiq commented:
@bors r-

llogiq commented:
@bors r+

bors commented:
:pushpin: Commit aaf04dc0431aaf4be9f3125dd9cfcb80f2ae7e25 has been approved by `llogiq`

<!-- @bors r=llogiq aaf04dc0431aaf4be9f3125dd9cfcb80f2ae7e25 -->
<!-- homu: {"type":"Approved","sha":"aaf04dc0431aaf4be9f3125dd9cfcb80f2ae7e25","approver":"llogiq"} -->

bors commented:
:hourglass: Testing commit aaf04dc0431aaf4be9f3125dd9cfcb80f2ae7e25 with merge fe7254ff6f9ed485fe3d8b31c97c0beac1102d12...
<!-- homu: {"type":"BuildStarted","head_sha":"aaf04dc0431aaf4be9f3125dd9cfcb80f2ae7e25","merge_sha":"fe7254ff6f9ed485fe3d8b31c97c0beac1102d12"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/5758271056?check_suite_focus=true), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/5758248483?check_suite_focus=true), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/5758449368?check_suite_focus=true)
Approved by: llogiq
Pushing fe7254ff6f9ed485fe3d8b31c97c0beac1102d12 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"llogiq","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/5758449368?check_suite_focus=true","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/5758271056?check_suite_focus=true","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/5758248483?check_suite_focus=true"},"merge_sha":"fe7254ff6f9ed485fe3d8b31c97c0beac1102d12"} -->
